src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -2198,19 +2198,54 @@ module Queries = struct
         Mina_lib.snark_pool mina |> Network_pool.Snark_pool.resource_pool
         |> Network_pool.Snark_pool.Resource_pool.all_completed_work )
 
+  (* Auxiliary function to fetch all pending work *)
+  let get_pending_work mina =
+    let snark_job_state = Mina_lib.snark_job_state mina in
+    let snark_pool = Mina_lib.snark_pool mina in
+    let fee_opt =
+      Mina_lib.(
+        Option.map (snark_worker_key mina) ~f:(fun _ -> snark_work_fee mina))
+    in
+    Work_selector.pending_work_statements ~snark_pool ~fee_opt snark_job_state
+
   let pending_snark_work =
     field "pendingSnarkWork" ~doc:"List of snark works that are yet to be done"
       ~args:Arg.[]
       ~typ:(non_null @@ list @@ non_null Types.pending_work)
-      ~resolve:(fun { ctx = mina; _ } () ->
-        let snark_job_state = Mina_lib.snark_job_state mina in
-        let snark_pool = Mina_lib.snark_pool mina in
-        let fee_opt =
-          Mina_lib.(
-            Option.map (snark_worker_key mina) ~f:(fun _ -> snark_work_fee mina))
+      ~resolve:(fun { ctx = mina; _ } () -> get_pending_work mina)
+
+  let pending_snark_work_range =
+    field "pendingSnarkWorkRange"
+      ~doc:
+        "Find any sequence of pending snark works between two indexes in the \
+         pending snark work pool"
+      ~args:
+        Arg.
+          [ arg "startingIndex"
+              ~doc:
+                "The first index to be taken from the pending snark work pool"
+              ~typ:(non_null Types.Input.UInt32.arg_typ)
+          ; arg' "endingIndex"
+              ~doc:
+                "The last index to be taken from the pending snark work pool \
+                 (included). If not specified, only the work at index \
+                 [startingIndex] will be returned. An empty list will be \
+                 returned if startingIndex > endingIndex"
+              ~typ:Types.Input.UInt32.arg_typ ~default:Unsigned.UInt32.max_int
```

Comment:
My personal preference, would for this to also be non-null.

Reply:
I would expect this working if `startingIndex = endingIndex = 0`, so I would not be in favor to expect `endingIndex ≠ 0`

Reply:
Curious what the behavior is when it has a default of max int. Will it be null only if a user explicitly specified "null"?

Reply:
I changed the default behavior. Now
- if `endingIndex ≥ length(pending_work)`, everything from `startingIndex` is returned. (`max_int` is still the default)
- `endingIndex` is excluded
- `[]` is returned when `startingIndex ≥ endingIndex` or `startingIndex ≥ length(pending_work)`

Reply:
I still thinking having max-int is unnecessary logic. I think we can come up with a solution that removes this and enforces the invariants we know to be true. For example

```ocaml
  let pending_snark_work_range =
    field "pendingSnarkWorkRange"
      ~doc:
        "Find any sequence of pending snark works between two indexes in the \
         pending snark work pool"
      ~args:
        Arg.
          [ arg "startingIndex"
              ~doc:
                "The first index to be taken from the pending snark work pool"
              ~typ:(non_null Types.Input.UInt32.arg_typ)
          ; arg "endingIndex"
              ~doc:
                "The last index to be taken from the pending snark work pool \
                 (excluded). If not specified or greater than the pending \
                 snark work list, all elements from index [startingIndex] will \
                 be returned. An empty list will be returned if \
                 [startingIndex, endingIndex] is not a valid range of the \
                 pending snark work list."
              ~typ:Types.Input.UInt32.arg_typ 
          ]
      ~typ:(non_null @@ list @@ non_null Types.pending_work)
      ~resolve:(fun { ctx = mina; _ } () start_idx end_idx ->
        let pending_work = get_pending_work mina in
        let pending_work_size = pending_work |> List.length |> Unsigned.UInt32.of_int in
        let less_than uint1 uint2 = Unsigned.UInt32.compare uint1 uint2 < 0  in 
        match end_idx with 
        | None ->
            (* drop handles case when start_idx is greater than pending work and is O(start_time)*)
            let start = Unsigned.UInt32.to_int start_idx in
            List.drop pending_work start
        | Some end_idx when  less_than start_idx  end_idx  && less_than end_idx pending_work_size ->
          let len = Unsigned.UInt32.sub end_idx start_idx in
          let pos = Unsigned.UInt32.to_int start_idx in
          let len = Unsigned.UInt32.to_int len in
          List.sub ~pos ~len pending_work
        | _ ->
            [] )


```

Reply:
Is there a reason for prefering UInt32 manipulation over to_int conversion from the start ?

Reply:
I didn't want to run any transformations on user controlled data before checking, and it was more convenient to write with the match case. You can do uint to int conversions safely, but I wasn't sure how the js one was implemented, so I just took the safe route without reading the implementation.

Reply:
I pushed a new version with your suggestion but I modified some match cases to match the spec.